apm-sniffer/apm-sdk-plugin/nats-client-plugin/src/main/java/org/apache/skywalking/apm/plugin/nats/client/WriterSendMessageBatchInterceptor.java

```diff
@@ -0,0 +1,121 @@
+/*
+ *   Licensed to the Apache Software Foundation (ASF) under one or more
+ *   contributor license agreements.  See the NOTICE file distributed with
+ *   this work for additional information regarding copyright ownership.
+ *   The ASF licenses this file to You under the Apache License, Version 2.0
+ *   (the "License"); you may not use this file except in compliance with
+ *   the License.  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.apache.skywalking.apm.plugin.nats.client;
+
+import io.nats.client.Connection;
+import io.nats.client.Message;
+import io.nats.client.impl.NatsMessage;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.ContextSnapshot;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+import static org.apache.skywalking.apm.plugin.nats.client.NatsCommons.addCommonTag;
+import static org.apache.skywalking.apm.plugin.nats.client.NatsCommons.injectCarrier;
+import static org.apache.skywalking.apm.plugin.nats.client.NatsCommons.skipTrace;
+
+public class WriterSendMessageBatchInterceptor implements InstanceMethodsAroundInterceptor {
+
+    private static final ILog LOGGER = LogManager.getLogger(NatsCommons.class);
+
+    private static final Field NEXT_FIELD;
+
+    static {
+        Field field;
+        try {
+            field = NatsMessage.class.getDeclaredField("next");
+            field.setAccessible(true);
+        } catch (NoSuchFieldException e) {
+            field = null;
+        }
+        NEXT_FIELD = field;
+    }
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+        String connection = ((Connection) objInst.getSkyWalkingDynamicField()).getConnectedUrl();
```

Comment:
If you hold this connection at the instance level, would this connection have been closed? If in that case, would `getConnectedUrl` fail?

As you only need the URL, how about just keeping the URL in the `SkyWalkingDynamicField`?

Reply:
` connection`  represent a cluster  , it include more than one url . it will reconnect another server if current server fail. 
`getConnectedUrl ` means get current useable url , so I couldn't  hold  URL

Reply:
If it is un-usable? What happens?

Reply:
If the library healthy check success , And write to network failed . SendMessage will error . We could record this error .
The error should associated with  current URL . It's right

Reply:
I mean, if there is no connected server, is the URL still correct? As well as the connection? In many libs, unconnected status makes the connection in closed status. Then you can't read things from it.

Reply:
`sendMessage` will block to get a useable url  util reconnect success

Reply:
But in your codes, you cached the connection in the dynamic field, when it is flushed, the before method in your interceptor executes before the real flush method.
This would be a tricky point.

Reply:
The library  quit the block first , then update currentServer . It's strange . 
 I will confirm it carefully .  fix it  and get really url

Reply:
> The library quit the block first , then update currentServer . It's strange .

What do you mean `block first`? I just give you a possible example to show you, `cache the connection` is tricky in many cases. You should be careful.

Reply:
If your `before method` is being blocked in these cases, people would report bugs. We should be careful to recheck.

Reply:
Conclusion :
1 . Get url from sendMessage arguments 
2. for `queue` method , I want to remove URL tag , as couldn't record real url .

Reply:
URL tag is not required but only recommended. 

BTW, endpoint name usually doesn't carry the `IP:port`, URL tag carries.

Reply:
OK

Reply:
Unfortunately , If I get URL from method argument , I have to use java reflect . Because `ip` `port` field is private in the argument . 
If get URL from `Connection ` , the URL maybe outdated . 
I'm a little torn.

Reply:
Is that connection class specific for Nats? Or a general class?

Reply:
`NatsConnection`  , only for Nats

Reply:
Then it is easy. You could instrument that class and use dynamic field updated with time.

Reply:
Good idea

Reply:
> Unfortunately , If I get URL from method argument , I have to use java reflect . Because `ip` `port` field is private in the argument .

On subscriber side , I still use  `NatsConnection#getConnectedUrl()` to get current url , although it maybe outdated.
Because  subscribe-method no any argument about server address .
Even if the I use dynamic field at someplace, I don't have  chance update the field  , as field-assignment

Reply:
If it is not correct, how about we remove the URL out of tag and operation name of span? Is the `peer` correct?

Reply:
1. If it's incorrect , we remove it ? I think the probability is low . Are you sure ? 
2.  only exitSpan need `peer` , entrySpan don't need ,  exitSpan  `peer` is right .

Reply:
Actually, all operation names usually don't carry the IP:port, I thought you are able to get one, so we have this discussion.

You could check Kafka sub/pub spans, both of them don't include the information. The `peer` is expected to be correct, at least for the exit span.

Reply:
I think I could  ,  but couldn't  .  Very sad . 
Let's stop this discussion

Reply:
Then, let's keep `peer` correct and remove IP:port out of operation. Keep operation name only carrying the NATS operation information.

Reply:
operation name  is ok .